utilities: fix issues related to conversion of lengths to pt

dixslyf opened the pull request:
This PR fixes some issues regarding the conversion of lengths to pt ~~in hopes of progressing #74~~. Please see each commit's message for details.

I don't see any regressions in `tablex-test.typ` (compiled it to PNG before and after this PR's commits, then checked their hashes).

~~I'm marking this as a draft for now as I haven't added any tests. I did some quick testing with negative line expansion, and surprisingly, it seems to work, so I'd like to investigate this further before readying this PR.~~

EDIT: should resolve #74.

PgBiel commented:
Oh, I didn't realize that the problem occurred even when you were reducing the stroke below a single cell. That is interesting. Thanks for looking into this!

PgBiel commented on src/utilities.typ:
```diff
@@ -114,6 +114,68 @@
+        panic("Cannot convert length to pt ('styles' not specified).")
+    }
+
+    // Convert the em part
+    //
+    // `len.em` is a float.
+    // We cannot simply do `len < 0pt` as that throws an error when
+    // `len` is compound (i.e., has both an absolute component and an em component).
+    // So, we have to handle the absolute and em components separately.
+    let em_len_in_pt = if len.em < 0 {
+        -measure(line(length: -len.em * 1em), styles).width
+    } else if len.em > 0 {
+        measure(line(length: len.em * 1em), styles).width
+    } else {
+        0pt
+    }
```
1. By the way, I'm working on removing all usages of snake_case from the codebase, so I fully understand that you wanted this to be coherent with the rest :p. But let's use kebab-case for all future variable names so I can skip this later during my refactor.
2. We can't use `.em` as that was only added in Typst 0.7.0 iirc (tablex aims to support 0.2.0 or later, although increasing the minimum version later isn't really out of question, but I'd prefer to avoid it for tablex 0.1.0 for now). We'll probably have to perform some regex matching on repr() for now (a bit ugly, yeah)...

dixslyf replied:
>     2. We can't use `.em` as that was only added in Typst 0.7.0 iirc (tablex aims to support 0.2.0 or later, although increasing the minimum version later isn't really out of question, but I'd prefer to avoid it for tablex 0.1.0 for now). We'll probably have to perform some regex matching on repr() for now (a bit ugly, yeah)...  (Edit: same remark goes for the use of `.ratio` and `.length` below)

Ahh, I was wondering why you weren't using those. Makes sense. I'll make some changes.

PgBiel commented on src/utilities.typ:
```diff
@@ -129,73 +191,16 @@
-        let len_per_frac = frac_total / frac_amount
-
-        (len_per_frac * (len / 1fr)) + 0pt
+        convert-fraction-type-topt(len, frac_amount, frac_total)
```
```suggestion
        convert-fraction-type-to-pt(len, frac_amount, frac_total)
```

dixslyf commented:
I've rewritten the functions to use regex (yeah, quite ugly) and added some tests.

One concern I have: the `repr` for the `abs` component of a length rounds to a precision of 2 ([src](https://github.com/typst/typst/blob/7bb4f6df44086b4c1120b227f7ae963e6c2ad5ab/crates/typst/src/geom/abs.rs#L138)), so there will be a loss of precision. This is true for at least Typst 0.9.0, but I haven't looked at previous versions and other length types.

PgBiel commented:
> One concern I have: the `repr` for the `abs` component of a length rounds to a precision of 2 ([src](https://github.com/typst/typst/blob/7bb4f6df44086b4c1120b227f7ae963e6c2ad5ab/crates/typst/src/geom/abs.rs#L138)), so there will be a loss of precision. This is true for at least Typst 0.9.0, but I haven't looked at previous versions and other length types.

That seems to be the case since before Typst was released... In this case, I think a good workaround might be to keep the old algorithm (just measure the line) if there are no negative numbers in the length's parts (checking if either of the "minus" symbols is present in the repr() should be enough). That way, we only lose a bit of precision when dealing with the specific case of negative numbers. Of course, this is bound to be improved if/when we switch to 0.7.0's constructs.

PgBiel commented on tablex-test.typ:
```diff
@@ -783,3 +783,84 @@ Combining em and pt (with a stroke object):
+            frac-amount: frac-amount,
+            frac-total: frac-total,
+        )
+
+        [
+            Input type: #type(len) \
+            Input length: #len \
+            Expected: #expected \
+            Actual: #actual \
+            Pass: #{expected == actual} \
+        ]
```
This should be a simple `assert` (compilation should fail if these tests fail):

```suggestion

        assert(expected == actual)
```

PgBiel commented on tablex-test.typ:
```diff
@@ -783,3 +783,84 @@ Combining em and pt (with a stroke object):
     [E]
 )
+
+= Length to pt conversion
```
Just for consistency with the other tests (might change them later)

```suggestion
*Length to pt conversion*
```

PgBiel commented on src/utilities.typ:
```diff
@@ -114,88 +119,124 @@
+    // for fractional lengths (e.g., .5em becomes "0.5em"). Negative parts will always
+    // have a single minus sign or hyphen (depending on the Typst version) in front without
+    // any separating space.
+    let len-repr = repr(len)
+
```
Re: my comment on using the old algorithm where possible (I believe, without any minus signs, it will always be positive, as it'd be weird to have text with negative size), the code below should be reasonable:

```suggestion
    let len-repr = repr(len)
    if "−" not in len-repr and "-" not in len-repr {
       // No way the length is negative here, so we just measure the line straight away.
       // When the length CAN be negative, measuring the line might not give accurate
       // results, so we perform some extra checks.
       return measure(line(length: len), styles).width
    }
```

PgBiel replied:
Actually, I've just considered that we only have to do this regex matching precisely when one of the two components (em / pt) is positive and the other negative. When both are negative, or there is only one and it is negative, you can just multiply the length by -1 and measure, then multiply back by -1. The case where both (or the only one) are positive is the one I propose above.

So perhaps a better solution is to keep track of:
1. If there's an em component (`"em" in len-repr`);
2. If there's a pt component (`"pt" in len-repr`);
3. Perhaps which ones are negative (you could just count the amount of minus characters in the string).

Then, when the amount of minus is 0 (code I suggest above), just measure normally; when the amount of minus matches the amount of components, measure the length multiplied by -1 then multiply the result by -1; and, otherwise, do the regex thing.

That way, we can reduce the need for the "possibly imprecise" approach as much as possible. Tell me what you think about this idea.

PgBiel replied:
Sketch:

```suggestion
    let len-repr = repr(len)
    let has-em = "em" in len-repr
    let has-pt = "pt" in len-repr
    if "−" not in len-repr and "-" not in len-repr {
       // No way the length is negative here, so we just measure the line straight away.
       // When the length CAN be negative, measuring the line might not give accurate
       // results, so we perform some extra checks.
       return measure(line(length: len), styles).width
    } else if not has-em or not has-pt or len-repr.matches(regex("−|-")).len() == 2 {
       // All components (em / pt) are negative, so we can just measure the whole length
       // multiplied by -1 and then multiply by -1 again.
       return -measure(line(length: -len), styles).width
    }
    // One component is negative and the other is positive, so we'll need to measure them
    // separately.
```

dixslyf replied:
That is a smart idea! This got me thinking: we might be able to avoid regex and the precision problem completely. If I remember correctly, drawing a line with a negative length and measuring it gives a length of 0pt. What we could do is:
- Draw a line with the given length, regardless of whether it's positive or negative.
- If the measured length is more than 0pt, then the given length must have been positive and we simply return the measured length.
- If not (i.e., the measured length is 0pt), the given length must either be 0pt or negative.
- To differentiate, we multiply the given length by -1, then draw and measure a line again.
- If measuring the line still gives 0pt, then the given length must have been 0pt, so we return 0pt.
- Otherwise, it must be negative, so we multiply the measured length by -1 and return it.

dixslyf replied:
I did some quick benchmarking using [hyperfine](https://github.com/sharkdp/hyperfine). I'm a rookie at benchmarking, but the results seem okay.

At 464ad1d8d4866909cee03765b6a71864a6d76fd5 (before):
```
Benchmark 1: typst compile tablex-test.typ
  Time (mean ± σ):      4.406 s ±  0.286 s    [User: 4.166 s, System: 0.239 s]
  Range (min … max):    4.101 s …  4.762 s    10 runs
```

At 162f18f3b40a265e19b848a4b7036dfeee167205 (after):
```
Benchmark 1: typst compile tablex-test.typ
  Time (mean ± σ):      4.276 s ±  0.213 s    [User: 4.063 s, System: 0.213 s]
  Range (min … max):    4.085 s …  4.564 s    10 runs
```

PgBiel commented on tablex-test.typ:
```diff
@@ -783,3 +783,84 @@ Combining em and pt (with a stroke object):
+#convert-length-to-pt-test(-0.5% + 0pt + 0em, -0.5pt) \
+#convert-length-to-pt-test(-0.5% + 0pt - 0.5em, -5.5pt) \
+#convert-length-to-pt-test(-0.5% - 0.5pt + 0em, -1pt) \
+#convert-length-to-pt-test(-0.5% - 0.5pt - 0.5em, -6pt)
```
If this truly tackles issue #74 , then feel free to add some tests for it as well :)

PgBiel reviewed (commented):
This is looking great! Thanks for the clean code :sunglasses:

dixslyf commented:
I've gone ahead and implemented what I came up with in https://github.com/PgBiel/typst-tablex/pull/84#discussion_r1413024123. I've also managed to come up with a solution for the same precision issue for `relative` lengths (we don't use `repr` for `fraction` and `ratio`, so those are unaffected).

On the testing side:
- Added tests for line expansion
- Updated the conversion tests to check the return type (`length`) and use values that can better ensure that there is no precision loss

I think we're nearly done with this PR. All that's left is another round of review and ensuring that this PR works with older versions of Typst.

I've left the commit history mostly intact because I didn't want to accidentally remove commits when force pushing. However, I can squash some of them if you prefer.

PgBiel commented on src/utilities.typ:
```diff
@@ -114,88 +119,124 @@
+// Convert a ratio type length to pt
+//
+// page-size: equivalent to 100%
+#let convert-ratio-type-to-pt(len, page-size) = {
```
Yeah, we won't ever pass non-`pt` to those functions. But indeed sounds like it'd be wise to add an `assert(type(page-size) == _length_type and "em" not in repr(page-size))` or something like that.

dixslyf commented:
Re-ordered the commits so that changes to `tablex-test.typ` come later. Makes it easier to check for regressions before `tablex-test.typ` got modified.

I managed to hack together a bunch of scripts using Nix and [diff-pdf](https://github.com/vslavik/diff-pdf) to semi-automate regression testing for the different Typst versions. (Didn't include them here since they're out of scope and pretty makeshift, but they can be found here: https://github.com/dixslyf/typst-tablex/commit/7d078ebff5f91010395103aa624af068686fcac8.) I don't see any regressions for any of the supported Typst versions.

PgBiel commented:
I plan on having a better testing system in the future. I'll probably consider using PNG instead of PDF though as it's a pain to work with :p

I'm particularly interested in Lau's efforts towards a standard testing solution for packages, although I might also try out something like CeTZ's system. We'll see...

dixslyf commented:
> I'll probably consider using PNG instead of PDF though as it's a pain to work with :p

Indeed, PDFs are a huge pain to work with. I initially wanted to go with PNG for my tests, but PNG export was only added in Typst v0.5.0. What I tried at first was to convert the PDFs to PNGs using ImageMagick (I also tried `pdftoppm`), but the conversion took a while and you need to tune the options to get decent quality output. I later realized you could just use `magick compare` on two PDFs (it'll do the conversion for you, but you'll still need to tune the options) to generate a diff image (the exit code also tells you whether the PDFs were similar). I went with `diff-pdf` (which does more or less the same thing) in the end just because it was easier to use.

Another thing I tried was to compare the output across different Typst versions (e.g., compare the outputs for Typst v0.2.0 and v0.9.0). The outputs were slightly different — mostly differences in spacing. However, even for pages that look the same to our eyes, if you compared them using `magick compare`, there would be differences in the anti-aliasing (I think) of rendered text. I figured it wasn't worth it to test across Typst versions.

> I'm particularly interested in Lau's efforts towards a standard testing solution for packages,

Same!

> although I might also try out something like CeTZ's system. We'll see...

I took a look at CeTZ's testing when I was trying to test this PR, and it did look nice and promising. My only concern was it tests only a single version of Typst as far as I can tell. It's probably not too difficult to expand on that, however (e.g., modifying `PATH` to use a different Typst version, though fetching the different Typst versions and caching them somewhere sounds like a bit of a pain).
